pkg/api/pod/util.go

```diff
@@ -672,19 +673,113 @@ func dropDisabledNodeInclusionPolicyFields(podSpec, oldPodSpec *api.PodSpec) {
 	}
 }
 
-// dropDisabledMatchLabelKeysField removes disabled fields from PodSpec related
-// to MatchLabelKeys only if it is not already used by the old spec.
-func dropDisabledMatchLabelKeysField(podSpec, oldPodSpec *api.PodSpec) {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpread) && !matchLabelKeysInUse(oldPodSpec) {
+// dropDisabledMatchLabelKeysFieldInPodAffinity removes disabled fields from PodSpec related
+// to MatchLabelKeys in required/preferred PodAffinity/PodAntiAffinity only if it is not already used by the old spec.
+func dropDisabledMatchLabelKeysFieldInPodAffinity(podSpec, oldPodSpec *api.PodSpec) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodAffinity) || podSpec == nil || podSpec.Affinity == nil {
+		return
+	}
+
+	if affinity := podSpec.Affinity.PodAffinity; affinity != nil {
+		if !matchLabelKeysInRequiredPodAffinityInUse(oldPodSpec) {
+			for i := range affinity.RequiredDuringSchedulingIgnoredDuringExecution {
+				podSpec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution[i].MatchLabelKeys = nil
+			}
+		}
+		if !matchLabelKeysInPreferredPodAffinityInUse(oldPodSpec) {
+			for i := range affinity.PreferredDuringSchedulingIgnoredDuringExecution {
+				podSpec.Affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution[i].PodAffinityTerm.MatchLabelKeys = nil
+			}
+		}
+	}
+
+	if affinity := podSpec.Affinity.PodAntiAffinity; affinity != nil {
+		if !matchLabelKeysInRequiredPodAntiAffinityInUse(oldPodSpec) {
+			for i := range podSpec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+				affinity.RequiredDuringSchedulingIgnoredDuringExecution[i].MatchLabelKeys = nil
+			}
+		}
+		if !matchLabelKeysInPreferredPodAntiAffinityInUse(oldPodSpec) {
+			for i := range podSpec.Affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
+				affinity.PreferredDuringSchedulingIgnoredDuringExecution[i].PodAffinityTerm.MatchLabelKeys = nil
+			}
+		}
+	}
+}
+
+// dropDisabledMatchLabelKeysFieldInTopologySpread removes disabled fields from PodSpec related
+// to MatchLabelKeys in TopologySpread only if it is not already used by the old spec.
+func dropDisabledMatchLabelKeysFieldInTopologySpread(podSpec, oldPodSpec *api.PodSpec) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpread) && !matchLabelKeysInTopologySpreadInUse(oldPodSpec) {
 		for i := range podSpec.TopologySpreadConstraints {
 			podSpec.TopologySpreadConstraints[i].MatchLabelKeys = nil
 		}
 	}
 }
 
-// matchLabelKeysInUse returns true if the pod spec is non-nil
+// matchLabelKeysInRequiredPodAffinityInUse returns true if the pod spec is non-nil
```

Comment:
The general guidance is that we don't need all this complexity. If the field is in use anywhere, it's fair game and we can keep adding more.

Just having a single `matchLabelKeysFieldInPodAffinity` is enough.

And that can lead to less test cases.